lib/fast_inserter/fast_inserter_base.rb

```diff
@@ -92,20 +95,44 @@ def fast_insert_group(group)
       end
     end
 
+    def existing_values_sql(group_of_values)
+      sql = "SELECT #{@variable_columns.join(', ')} FROM #{@table_name} WHERE #{values_hash_to_sql(@static_columns)}"
+
+      # Using the database to filter existing values is not enabled by default since large IN queries in the SQL
+      # statement can be bad for database performance.
+      if @options[:filter_existing_in_database]
+        if @variable_columns.length > 1
+          group_of_values_sql = group_of_values.map do |group|
+            values_hash = variable_column_values_to_hash(group)
+            "(#{values_hash_to_sql(values_hash)})"
+          end.join(' OR ')
+
+          sql += " AND (#{group_of_values_sql})"
+        else
+          values_to_check = ActiveRecord::Base.send(:sanitize_sql_array, ['?', group_of_values.flatten])
+          sql += " AND #{@variable_columns.first} IN (#{values_to_check})"
+        end
+      end
+
+      sql
+    end
+
     # Queries for the existing values for a given group of values
     def existing_values(group_of_values)
-      sql = "SELECT #{@variable_columns.join(', ')} FROM #{@table_name} WHERE #{existing_values_static_columns}"
+      sql = existing_values_sql(group_of_values)
```

Comment:
what do you think about breaking apart the two logical paths of a) existing values via ruby b) existing values via sql into two separate methods fully?

during review, i found it hard to review `fast_insert_group` ->  `existing_values` -> `existing_values_sql` -> more code in `existing_values`.

especially i think it was difficult since in `existing_values_sql ` we check `if @options[:filter_existing_in_database]` and have forking logic based on it.

or, maybe some comments outlining what is happening in more detail. this code is getting pretty complex 😅 . for example comment on what the `existing_values` method does, what the `existing_values_sql` method does, and why they are both needed. i wonder if `existing_values_sql` just needs a better name now?